### PR TITLE
Mate.hw

### PR DESCRIPTION
**Pull request by YeaaaaH**
Home work01

### CODE REVIEW

File: CalcApp/pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <parent>
         <artifactId>junior</artifactId>
-        <groupId>ru.job4j</groupId>
+        <groupId>Shtef</groupId>
```
Comment (Siril-O) [inline CalcApp/pom.xml]: Why do you have 2 pom files?

File: HW/pom.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>junior</artifactId>
```
Comment (Siril-O) [inline HW/pom.xml]: use some descriptive package structure ex: com.ua.junior

Comment (Siril-O) [reply HW/pom.xml]: https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

File: HW/src/main/java/hw01/HomeWork01.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
+package hw01;
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: use more desctiptive package name

File: HW/src/main/java/hw01/HomeWork01.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+import java.util.Objects;
+
+/**
+ * Home work #1.
+ */
+public class HomeWork01 implements Cloneable {
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: Use some other name please

File: HW/src/main/java/hw01/HomeWork01.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    /**
+     * Taks #1.
+     * @param a first team scores
+     * @param b second team scores
+     * @param c third team scores
+     * @param d fourth team scores
+     * @return 2 - if completely rigth, 1 - if only winner, 0 - other cases
+     */
+    public int bets(final int a, final int b, final int c, final int d) {
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: this are separate tasks. Class should be responsible for single task or set of related tasks. Create multiple classes for each task .

File: HW/src/main/java/hw01/HomeWork01.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+        if (a < 0 || b < 0 || c < 0 || d < 0) {
+            System.out.println("You put numbers less than 0, try again");
+            return 0;
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: 0 is valid response, here you should return error response like -1

File: HW/src/main/java/hw01/HomeWork01.java (hunk 5 above)
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: not a valid variables names use something self descriptive like firstTeamScore

File: HW/src/main/java/hw01/HomeWork01.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            return (a == c && b == d) ? 2
+                : (((a > b && c > d) || (b > a && d > b)) ? 1
+                        : 0);
+    }
+    /**
+     * Draw a rhombus.
+     * @param n length
+     */
+    public void drawRhombus(final int n) {
+        int x = n / 2 - 1;
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: what does x means ? Give variables descriptive names

File: HW/src/main/java/hw01/HomeWork01.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < n; j++) {
+                if (i == n / 2 || j == n / 2) {
+                    System.out.print("*");
+                } else if (j == n / 2 - i || j == n / 2 + i || j == i - n / 2 || j == n - i + n / 2 - 1) {
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: This is unreadable, try to make it more understandable, maybe extract some variables with descriptive names

File: HW/src/main/java/hw01/HomeWork01.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        System.out.println();
+    }
+    /**
+     * Powing.
+     * @param number number for calculating
+     * @param n needed pow
+     * @return number^n
+     */
+    public int powImpl(int number, int n) {
+        if (n == 0) {
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: What if number is -10 ?

File: HW/src/main/java/hw01/HomeWork01.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+            number *= number1;
+        }
+        return number;
+    }
+    /**
+     * Factorial recursion.
+     * @param n number for calculating
+     * @return n!
+     */
+    public int fact(final int n) {
```
Comment (Siril-O) [inline HW/src/main/java/hw01/HomeWork01.java]: use full name not fact

**Comment by Siril-O** [comment]
Home work 2 should be committed to separate branch and you should create separate pull request. Cherry pick your commit to another branch and create one more pull request

File: CalcApp/pom.xml (hunk 1 above)
Comment (YeaaaaH) [reply CalcApp/pom.xml]: For really I don't even know. Created 1 for main pom, and one for each chapter, what is a right approach?

File: HW/src/main/java/hw01/HomeWork01.java (hunk 3 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 4 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 5 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 6 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 5 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 7 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 8 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 9 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/hw01/HomeWork01.java (hunk 10 above)
Comment (YeaaaaH) [reply HW/src/main/java/hw01/HomeWork01.java]: done

File: HW/src/main/java/home_work_01/MainHW.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        this.firstName = firstName;
+        this.lastName = lastName;
+    }
+    /**
+     * Object class deep copy method
+     * @return
+     * @throws CloneNotSupportedException
+     */
+    @Override
+    protected Object clone() throws CloneNotSupportedException {
```
Comment (Siril-O) [inline HW/src/main/java/home_work_01/MainHW.java]: better return  MainHW  here

**Pull request merged**